File: backend/src/main/java/com/resolvo/backend/notification/NotificationService.java
package com.resolvo.backend.notification;

import com.resolvo.backend.common.dto.PageResponse;
import com.resolvo.backend.exception.ResourceNotFoundException;
import com.resolvo.backend.exception.UnauthorizedAccessException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@RequiredArgsConstructor
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Transactional
    public Notification createNotification(Long userId, String title, String message, NotificationType type, Long referenceId) {
        Notification notification = Notification.builder()
                .userId(userId)
                .title(title)
                .message(message)
                .type(type)
                .referenceId(referenceId)
                .read(false)
                .build();
        Notification saved = notificationRepository.save(notification);
        log.debug("Notification created: id={}, user={}, type={}", saved.getId(), userId, type);
        return saved;
    }

    @Transactional(readOnly = true)
    public PageResponse<Notification> getMyNotifications(Long userId, Pageable pageable) {
        Page<Notification> page = notificationRepository.findByUserIdOrderByCreatedAtDesc(userId, pageable);
        return new PageResponse<>(page);
    }

    @Transactional(readOnly = true)
    public long getUnreadCount(Long userId) {
        return notificationRepository.countByUserIdAndReadFalse(userId);
    }

    @Transactional
    public void markAsRead(Long notificationId, Long userId) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new ResourceNotFoundException("Notification not found"));
        if (!notification.getUserId().equals(userId)) {
            throw new UnauthorizedAccessException("Cannot modify another user's notifications");
        }
        notification.setRead(true);
        notificationRepository.save(notification);
    }

    @Transactional
    public void markAllAsRead(Long userId) {
        notificationRepository.markAllAsReadForUser(userId);
    }
}

File: frontend/src/services/notificationService.ts
import { apiClient } from '@/api/axiosClient'
import { NOTIFICATION_PATHS } from '@/constants/apiPaths'
import type { ApiResponse, PageResponse } from '@/types/api'
import type { Notification } from '@/types/notification'

export const notificationService = {
  getMyNotifications: async (page: number, size = 15): Promise<PageResponse<Notification>> => {
    const { data } = await apiClient.get<ApiResponse<PageResponse<Notification>>>(
      NOTIFICATION_PATHS.base,
      { params: { page, size } },
    )
    return data.data
  },

  getUnreadCount: async (): Promise<number> => {
    const { data } = await apiClient.get<ApiResponse<number>>(
      `${NOTIFICATION_PATHS.base}${NOTIFICATION_PATHS.unreadCount}`,
    )
    return data.data
  },

  markAsRead: async (id: number): Promise<void> => {
    await apiClient.patch<ApiResponse<void>>(
      `${NOTIFICATION_PATHS.base}${NOTIFICATION_PATHS.read(id)}`,
    )
  },

  markAllAsRead: async (): Promise<void> => {
    await apiClient.patch<ApiResponse<void>>(
      `${NOTIFICATION_PATHS.base}${NOTIFICATION_PATHS.readAll}`,
    )
  },
}
